### ISSUE TITLE
distinguish unset and empty values for storageClassName

### ISSUE THREAD

**Issue by bryanlarsen** [kubernetes/helm#2600]
In a Kubernetes 1.6 PVC, storageClassName has different behaviour when passed an empty string versus not specified.  An empty string means disable dynamic provisioning, not specified means use the default provisioner.

Charts should enable this behaviour, but I can't figure out how to make it happen using the go template syntax.  go template's if treats the empty string as falsy, and everything i try to use to distinguish an empty string either works the same on an undefined value (len), or crashes (eq nil).

Perhaps this is why Hugo added the 'isset' function.    Something like this would likely work if we had isset:

```
  {{- if (isset .Values.alertmanager.persistentVolume "storageClass") }}
  storageClassName: "{{ .Values.alertmanager.persistentVolume.storageClass }}"
  {{- end }}
```

Is there any way to do this?  If not can we pull in Hugo's isset or some other way to do it?

**Comment by seh** [comment]
While I agree that `isset` is a nicer solution, one workaround to consider is to honor sentinel storage class name that's invalid. [A valid name must be a DNS subdomain](https://github.com/kubernetes/kubernetes/blob/870585e8e1ca1133a67d4d44bfd4a9722b2b93c7/pkg/api/validation/validation.go#L279-L282). Following that down into _kubernetes/apimachinery_, we find [the governing regular expression](https://github.com/kubernetes/apimachinery/blob/master/pkg/util/validation/validation.go#L111), which is a lowercase letter or digit, followed by any number of lowercase letters, digits, or hyphens, followed by a lowercase letter or digit. In other words, hyphens are only allowed in the middle. Hence, the following are invalid storage class names:
- _-default_
- _default-_
- _!default_

You can imagine many more. What if you chose one of them and said that that's the value an release uses to choose the default provisioner?

**Comment by bryanlarsen** [comment]
That's a very reasonable solution for my own charts, but it's fairly hacky to get adopted as the right way to do things for public charts.   I'll try a PR for the prometheus chart using:

```
{{- if .Values.alertmanager.persistentVolume.storageClass }}
{{- if (eq "-" .Values.alertmanager.persistentVolume.storageClass) }}
  storageClassName: ""
{{- else }}
  storageClassName: "{{ .Values.alertmanager.persistentVolume.storageClass }}"
{{- end }}
{{- end }}
```

**Comment by seh** [comment]
Yes, Bryan, I wasn't trying to knock your proposed enhancement. I was just trying to help you make progress in the meantime.

**Comment by thomastaylor312** [comment]
Thanks for the request here @bryanlarsen. I'll tag it as a feature. If you wanted to submit something for this, please let us know!

**Comment by technosophos** [comment]
A PR on Sprig would be the best way to do this. I'll be rolling a new Sprig release for the upcoming Helm update.
